src/segment_reader.rs

```diff
@@ -342,6 +346,157 @@ impl AsyncSegmentReaderImpl {
     }
 }
 
+/// A wrapper around the AsyncSegmentReader.
+///
+/// It maintains a buffer and prefetches data to fill that buffer in the background so that
+/// every read call can read from the local buffer instead of waiting for server response.
+pub(crate) struct AsyncSegmentReaderWrapper {
+    buffer: Vec<u8>,
+    reader: Arc<Box<dyn AsyncSegmentReader>>,
+    read_size: usize,
+    pub(crate) offset: i64,
+    end_of_segment: bool,
+    is_truncated: bool,
+    outstanding: Outstanding,
+    handle: Handle,
+}
+
+impl AsyncSegmentReaderWrapper {
+    pub(crate) fn new(
+        handle: Handle,
+        reader: Arc<Box<dyn AsyncSegmentReader>>,
+        offset: i64,
+        buffer_size: usize,
+    ) -> Self {
+        let mut wrapper = AsyncSegmentReaderWrapper {
+            buffer: Vec::with_capacity(buffer_size),
+            reader,
+            read_size: buffer_size,
+            offset,
+            handle,
+            end_of_segment: false,
+            is_truncated: false,
+            outstanding: Outstanding {
+                end_of_segment: false,
+                cmd: None,
+                receiver: None,
+            },
+        };
+        wrapper.issue_request_if_needed();
+        wrapper
+    }
+    pub(crate) async fn read(&mut self, buf: &mut [u8]) -> StdResult<usize, ReaderError> {
+        self.issue_request_if_needed();
+        while self.buffer.is_empty() {
+            if self.outstanding.is_empty() {
+                return Ok(0);
+            }
```

Comment:
I feel like this is mixing paradigms. 
It's an async method. I think it's OK for it to return a result only after the operation is done.
So it probably makes more sense to only return once the requested data is available.